include/aws/common/cbor.h

```diff
@@ -0,0 +1,450 @@
+};
+
+struct aws_cbor_encoder;
+struct aws_cbor_decoder;
+
+/*******************************************************************************
+ * ENCODE
+ ******************************************************************************/
+
+/**
+ * @brief Create a new cbor encoder. Creating a encoder with a temporay buffer.
+ * Every aws_cbor_encoder_write_* will encode directly into the buffer to follow the encoded data.
+ *
+ * @param allocator
+ * @param initial_buffer Optional. the buffer for encoder to encode into, if not provided, the encoder will create a new
+ *      one.
+ * @return aws_cbor_encoder
+ */
+AWS_COMMON_API
+struct aws_cbor_encoder *aws_cbor_encoder_new(struct aws_allocator *allocator, struct aws_byte_buf *initial_buffer);
+
+AWS_COMMON_API
+struct aws_cbor_encoder *aws_cbor_encoder_release(struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Get the current encoded buffer from encoder.
+ *
+ * @param encoder
+ * @return struct aws_byte_cursor from the encoder buffer.
+ */
+AWS_COMMON_API
+struct aws_byte_cursor aws_cbor_encoder_get_encoded_data(const struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Clear the current encoded buffer from encoder.
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_reset_encoded_data(struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_UINT value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ *
+ * TODO: maybe add a width of the encoded value.
+ *
+ * @param encoder
+ * @param value value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_uint(struct aws_cbor_encoder *encoder, uint64_t value);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_NEGINT value to "smallest possible" in encoder's buffer.
+ * It represents (-1 - value).
+ *  Referring to RFC8949 section 4.2.1
+ *
+ *
+ * @param encoder
+ * @param value The argument to encode to negative integer, which is (-1 - expected_val)
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_negint(struct aws_cbor_encoder *encoder, uint64_t value);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_DOUBLE value to "smallest possible", but will not be encoded into half-precision float,
+ * as it's not well supported cross languages.
+ *
+ * To be more specific, it will be encoded into integer/negative/float
+ * (Order with priority) when the conversation will not cause precision loss.
+ *
+ * @param encoder
+ * @param value value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_double(struct aws_cbor_encoder *encoder, double value);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_BYTESTRING value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1, the length of "from" will be encoded first and then the value of "from" will
+ * be followed.
+ *
+ * @param encoder
+ * @param from value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_bytes(struct aws_cbor_encoder *encoder, const struct aws_byte_cursor from);
```

Comment:
trivial weird that the enum is BYTESTRING but this function is write_bytes(). Keep it consistent

BYTESTRING / write_bytestring()
or
BYTES / write_bytes()